server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java

```diff
@@ -114,4 +136,14 @@ protected void appendThreadPoolExecutorDetails(StringBuilder sb) {
             .append(", ");
     }
 
+    /**
+     * Returns the set of currently running tasks and their start timestamp.
+     * <p>
+     * Note that it is possible for a task that has just finished execution to be temporarily both in the returned map, and its total
+     * execution time to be included in the return value of {@code getTotalTaskExecutionTime()}. However, it is guaranteed that the
+     * task is reflected in at least one of those two values.
+     */
+    public Map<Runnable, Long> getOngoingTasks() {
```

Comment:
Updating the ongoing tasks start time doesn't seem right to me, since it just records the start time. At the end, in `AverageWriteLoadSampler` we do cap the time that we estimate for each task and for all ongoing tasks per sampling period.

Reply:
As discussed, the double counting scenario is a known side-effect of not synchronizing the total execution time and ongoing tasks, but for the estimation we're ok with it, due to the capping. If we find this at some point not accurate enough, we could address it by incurring some extra overhead , e.g. using the approach suggested by Francisco.